### PR TITLE
Fix issue 6227 - Deprecate comparison of different enums (revamp)

### PR DESCRIPTION
**Pull request by UplinkCoder**
This is a rebased version of #6444.
@andralex ready for merge

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@WalterBright ready for merge

File: test/runnable/gdb1.d [hunk 1]
```diff
@@ -8,7 +8,7 @@ r ARG1 ARG2
 echo RESULT=
 p args
 ---
-GDB_MATCH: RESULT=.*ARG1.*ARG2
+
```
Comment (wilzbach) [inline test/runnable/gdb1.d]: Why is this change needed?

**Review by wilzbach — COMMENTED** [review]
LGTM otherwise.
Is there a pattern in DMD to make scheduled deprecations easily greppable (like e.g. Phobos uses)?
Also I think this needs a PR to https://dlang.org/deprecate.html

File: test/runnable/gdb1.d (hunk 1 above)
Comment (UplinkCoder) [inline test/runnable/gdb1.d]: not sure ... it was in the original

File: changelog/b6227.dd [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Comparison of values belonging to different enums is deprecated.
+
+Since the comparison is done by comparing the value of the enumeration members
+we give up on the type-safety of the enumeration type possibly hiding a whole
+class of bugs that could be caught at compile time.
+
+This change doesn't affect the anonymous enumerations.
```
Comment (andralex) [inline changelog/b6227.dd]: s/the anonymous/anonymous/

File: changelog/b6227.dd [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+---
+enum Status
+{
+    GOOD,
+    BAD
+}
+enum OtherStatus
+{
+    OK
+    NO,
```
Comment (andralex) [inline changelog/b6227.dd]: We should follow the usual convention here, i.e. enum values in lowercase. Great example btw.

File: src/ddmd/expression.d [hunk 4]
```diff
@@ -15682,6 +15682,14 @@ extern (C++) final class EqualExp : BinExp
         if (e1.op == TOKtype || e2.op == TOKtype)
             return incompatibleTypes();
 
+        {
+            auto t1 = e1.type;
+            auto t2 = e2.type;
+            if (t1.ty == Tenum && t2.ty == Tenum && !t1.equivalent(t2))
+                deprecation("Comparison between different enumeration types (got %s and %s)",
```
Comment (andralex) [inline src/ddmd/expression.d]: Please suggest using `std.conv.asOriginalType` in the error message as a means to port code forward if needed.

**Review by andralex — CHANGES_REQUESTED** [review]
Almost there!

File: src/ddmd/expression.d (hunk 4 above)
Comment (UplinkCoder) [inline src/ddmd/expression.d]: I'd rather suggest `cast(long)`.
std.conv.asOriginalType will instantiate at-least three templates.

**Comment by UplinkCoder** [comment]
@andralex addressed your concernsm please review.

**Comment by UplinkCoder** [comment]
@wilzbach you will see the deprecations if you compile phobos.

**Review by andralex — APPROVED** [review]
thx!

**Comment by wilzbach** [comment]
> could you do the dlang pr ?
> My website foo is long gone.

It's not that hard, just open `deprecate.dd` and look how the other deprecations are done ;-)
This time I was nice though, please try to think about the dlang.org PRs in the future _before_ merging...
https://github.com/dlang/dlang.org/pull/1673/files

File: changelog/b6227.dd [hunk 5]
```diff
@@ -1,24 +1,21 @@
-Comparison of values belonging to different enums is deprecated.
+Comparison that takes into account the enumerated type improves code correctness by
+disallowing comparison of unrelated enumerated types.
```
Comment (CyberShadow) [inline changelog/b6227.dd]: Trailing whitespace. Also can changelog titles be several lines? README says no @wilzbach

Comment (CyberShadow) [inline changelog/b6227.dd]: Source code says no either.

Comment (CyberShadow) [inline changelog/b6227.dd]: Honestly I don't even understand what this sentence is supposed to mean.

Comment (wilzbach) [inline changelog/b6227.dd]: Yes it can't be and we should trigger a hard failure so that people, who don't read the README, are warned by the CI.

Also it seems that reviewers don't look at the changelog preview either as there's currently a missing parentheses:

http://dlang.org/changelog/2.075.0_pre.html

The Ddoc macro wouldn't be visible otherwise...
